bin/service-mango-health/src/processors/persister.rs

```diff
@@ -0,0 +1,392 @@
+            let ts: chrono::NaiveDateTime = row.get(1);
+            let ts_utc = chrono::DateTime::<Utc>::from_naive_utc_and_offset(ts, Utc);
+            let hr: Option<f64> = row.get(2);
+            result.insert(
+                key,
+                PersistedData {
+                    computed_at: ts_utc,
+                    health_ratio: hr.map(|x| I80F48::wrapping_from_num(x)), // TODO FAS What conversion should we use there ?
+                },
+            );
+        }
+
+        Ok(result)
+    }
+
+    async fn persist(
+        client: &mut Client,
+        previous: &mut HashMap<Pubkey, PersistedData>,
+        event: HealthEvent,
+    ) -> anyhow::Result<()> {
+        let mut updates = HashMap::new();
+
+        for component in &event.components {
+            if !Self::should_insert(&previous, event.computed_at, component.clone()) {
+                continue;
+            }
+
+            updates.insert(
+                component.account,
+                PersistedData {
+                    computed_at: event.computed_at,
+                    health_ratio: component.health_ratio,
+                },
+            );
+        }
+
+        let tx = client.transaction().await?;
+        Self::insert_history(&tx, &updates).await?;
+        Self::update_current(&tx, &updates, &previous).await?; // <- updated & present in previous
+        Self::insert_current(&tx, &updates, &previous).await?; // <- updated & not present in previous
+        Self::delete_old_history(&tx, event.computed_at).await?;
+        tx.commit().await?;
+
+        for (k, v) in updates {
+            previous.insert(k, v);
+        }
+
+        Ok(())
+    }
+
+    async fn insert_history<'tx>(
+        client: &Transaction<'tx>,
+        updates: &HashMap<Pubkey, PersistedData>,
+    ) -> anyhow::Result<()> {
+        for (key, value) in updates {
+            let key = key.to_string();
+            let ts = value.computed_at.naive_utc();
+            let hr = value.health_ratio.map(|x| x.to_num::<f64>());
+            let query = postgres_query::query!("INSERT INTO mango_monitoring.health_history (Pubkey, Timestamp, HealthRatio) VALUES ($key, $ts, $hr)", key, ts, hr);
+            query.execute(client).await.expect("Insertion failed");
+        }
+
+        Ok(())
+    }
+
+    async fn update_current<'tx>(
+        client: &Transaction<'tx>,
+        updates: &HashMap<Pubkey, PersistedData>,
+        previous: &HashMap<Pubkey, PersistedData>,
+    ) -> anyhow::Result<()> {
+        let to_update = updates
+            .into_iter()
+            .filter(|(key, _)| previous.contains_key(&key));
+
+        for (key, value) in to_update {
+            let key = key.to_string();
+            let ts = value.computed_at.naive_utc();
+            let hr = value.health_ratio.map(|x| x.to_num::<f64>());
+            let query = postgres_query::query!("UPDATE mango_monitoring.health_current SET Timestamp=$ts, HealthRatio=$hr WHERE Pubkey = $key", key, ts, hr);
```

Comment:
1. Did you intentionally split INSERT and UPDATE? In my experience INSERT ON CONFLICT UPDATE works well.
2. Having to insert all this data a second time sounds expensive. Maybe health_current could be a materialized view of health_history? With the right index finding the most recent timestamp data for each pubkey should be fast?